sympy/solvers/solvers.py

```diff
@@ -1557,7 +1557,22 @@ def _expand(p):
                         try:
                             t = Dummy('t')
                             iv = _solve(u - t, symbol, **flags)
+                            temp_soln = soln
                             soln = list(ordered({i.subs(t, s) for i in iv for s in soln}))
+                            count_nan = 0
+                            if S.NaN in soln:
+                                count_nan = soln.count(S.NaN)
+                                # try getting a solution for the equation
+                            if count_nan != 0:
+                                # heuristic -01(check whether the root satisfies the equation)
+                                index_list = []
+                                for index, s in enumerate(soln):
+                                    if s == S.NaN:
+                                        index_list.append(index)
+                                for i in index_list:
+                                    if checksol(f, symbol, temp_soln[i]):
```

Comment:
@asmeurer adding to what @valglad said about how it is able to extract `1`.
`index_list` keeps a track of the `indexes` of `nan` present, because the root present in that particular index number of  `temp_soln` caused the `nan`. Further checking for that root is done by `checksol`.